debug: thread_analyzer: add thread analyzer option to reset frame

dkalowsk opened the pull request:
Introduce the THREAD_ANALYZER_LONG_FRAME_PER_INTERVAL Kconfig to reset the longest frame value to zero after printing enabling analysis where the longest frame can now represents the most recent interval.  The longest overall frame is often at startup and this modification allows one to observe subsequent long frames.

ubieda commented on kernel/usage.c:
```diff
@@ -170,6 +170,13 @@ void z_sched_cpu_usage(uint8_t cpu_id, struct k_thread_runtime_stats *stats)
 }
 #endif /* CONFIG_SCHED_THREAD_USAGE_ALL */
 
+void z_sched_thread_reset_long_frame(struct k_thread *thread)
```
Shouldn't this be called from somewhere to clear the long-usage statistics after printing?

dkalowsk replied:
You are correct.  Missed that piece in moving from internal.  Thanks for catching it!

jeppenodgaard commented on kernel/usage.c:
```diff
@@ -170,6 +170,13 @@ void z_sched_cpu_usage(uint8_t cpu_id, struct k_thread_runtime_stats *stats)
 }
 #endif /* CONFIG_SCHED_THREAD_USAGE_ALL */
 
+void z_sched_thread_reset_long_frame(struct k_thread *thread)
```
Consider `inline` since it is a one line function.

jeppenodgaard replied:
Read up on https://www.kernel.org/doc/html/latest/process/coding-style.html#the-inline-disease and you are right.
Sorry for the noise.

jeppenodgaard commented on subsys/debug/thread_analyzer/thread_analyzer.c:
```diff
@@ -170,6 +171,11 @@ static void thread_analyze_cb(const struct k_thread *cthread, void *user_data)
 
 	cb(&info);
+
+#ifdef CONFIG_THREAD_ANALYZER_LONG_FRAME_PER_INTERVAL
```
Looks like this is already checked inside the function, so should not be necessary and will probably be optimized away anyway.

jeppenodgaard replied:
I would prefer less `ifdef` over saving space when optimization is not enabled.
This is a nit. Leave it as is and resolve if you wish.

jeppenodgaard replied:
From the coding style:
https://www.kernel.org/doc/html/latest/process/coding-style.html#conditional-compilation

jeppenodgaard replied:
As I understand it `#ifdef` inside `void z_sched_thread_reset_long_frame(struct k_thread *thread)` should not be used instead the function should be marked with `__maybe_unused` so the compiler will not complain if it is unused.

jeppenodgaard reviewed (commented):
LGTM

nashif commented on subsys/debug/thread_analyzer/thread_analyzer.c:
```diff
@@ -11,6 +11,7 @@
 #include <zephyr/kernel.h>
 /* For z_stack_space_get() */
 #include <kernel_internal.h>
+#include <ksched.h>
```
any inclusion of this file outside of the kernel/ is technical_debt++, this probably need to be implemented as part of k_thread_runtime_stats API instead and not using private function declarations. 

@peter-mitsis  please take a look at this and suggest an alternative solution.

peter-mitsis replied:
As I understand it, the inclusion of ksched.h is so that we can get access to the newly proposed z_sched_thread_reset_long_frame(). I'm OK for giving this a public name such as k_sched_thread_reset_longest_frame(). Someone is likely going to want that capability besides the thread_analyzer.

dkalowsk replied:
so is this moving something you'd want us to do as part of the PR?  Or are we holding until that move can be completed?

dkalowsk replied:
alright I believe this is now done.  Please take a look

nashif replied:
ok, so now we are including internal kernel headers and use the not-so-welcome newly introduced z_ prefixed functions, still not great. @peter-mitsis  can't this be converted to a public API? to make things a bit future proof?

dkalowsk committed:
debug: thread_analyzer: add thread analyzer option to reset frame

Use THREAD_ANALYZER_LONG_FRAME_PER_INTERVAL config to reset
longest frame value to zero after printing so that longest frame
now represents longest frame for the most recent interval.  The
longest overall frame is often at startup and this modification
allows one to observe subsequent long frames.

Signed-off-by: Jay Lawlor <jlawlor@amperecomputing.com>
Signed-off-by: Dan Kalowsky <dkalowsky@amperecomputing.com>

peter-mitsis reviewed (commented):
I think that we are almost there--just a couple of slight tweaks.

dkalowsk commented on include/zephyr/kernel.h:
```diff
@@ -103,6 +103,27 @@ typedef int (*_poller_cb_t)(struct k_poll_event *event, uint32_t state);
+static inline void
+	k_thread_runtime_stats_longest_frame_reset(__maybe_unused struct k_thread *thread)
+{
+#ifdef CONFIG_THREAD_ANALYZER_LONG_FRAME_PER_INTERVAL
```
Done.

dkalowsk commented on subsys/debug/thread_analyzer/thread_analyzer.c:
```diff
@@ -170,6 +170,9 @@ static void thread_analyze_cb(const struct k_thread *cthread, void *user_data)
 
 	cb(&info);
+
+	k_thread_runtime_stats_longest_frame_reset(thread);
```
Done.
